---
title: "Heterozygosity"
author: "Coline Caillon"
date: '2025-04-30'
output: github_document

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Jon's code: 
https://github.com/The-Eastern-Oyster-Genome-Project/2024_Eastern_Oyster_Population_Genomics/blob/main/Oyster_Genome_Population_Genomic_Analysis.md#heterozygosity

# Set up working directory and create symbolic link to the compressed VCF file

```{bash}
cd /home/ccaillon/
mkdir ROD_CADO_CC_working
cd ROD_CADO_CC_working
ln -s /home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz filtered.vcf.gz
```

```{bash}
# create popmap file with sample and treatment names
cd /home/ccaillon/ROD_CADO_CC_working
mkdir popmap_files
cd popmap_files
cp /home/Shared_Data/ROD_CADO/analysis/popmap treat_popmap
#add in treatment names to popmap file (w/ code)
head treat_popmap 
```

#Create treatment specific files containing a single column of all of the sample names within that treatment
```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/popmap_files

#Rename treatments
awk '$2 ~ /^C[123]$/ { $2 = "CONCON" } 
$2 ~ /^R[123]$/ { $2 = "CONROD" }
$2 ~ /^CH[123]$/ { $2 = "STRCON" }
$2 ~ /^RH[123]$/ { $2 = "STRROD" }
{ print }' treat_popmap > treat_popmap_renamed

awk '$2 == "CONCON" {print $1}' treat_popmap_renamed > CONCON.txt
awk '$2 == "STRCON" {print $1}' treat_popmap_renamed > STRCON.txt
awk '$2 == "CONROD" {print $1}' treat_popmap_renamed > CONROD.txt
awk '$2 == "STRROD" {print $1}' treat_popmap_renamed > STRROD.txt
```

#Subset the VCF by treatment and calculate heterozygosity for each treatment
```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/
mkdir -p treat_heterozygosity  # -p prevents error if directory exists

# CONCON
samples=$(cat popmap_files/CONCON.txt)
INDS=$(bcftools query -l ./filtered.vcf.gz | awk -v samples="$samples" '
BEGIN { split(samples, arr, "\n"); for (i in arr) sample_map[arr[i]] }
  { if ($0 in sample_map) printf "%s,", NR-1 }' | sed 's/,$//')
popStats -y GT --file <(bcftools view --threads 40 ./filtered.vcf.gz) --target "$INDS" > ./treat_heterozygosity/CONCON.het

# CONROD
samples=$(cat popmap_files/CONROD.txt)
INDS=$(bcftools query -l ./filtered.vcf.gz | awk -v samples="$samples" '
BEGIN { split(samples, arr, "\n"); for (i in arr) sample_map[arr[i]] }
  { if ($0 in sample_map) printf "%s,", NR-1 }' | sed 's/,$//')
popStats -y GT --file <(bcftools view --threads 40 ./filtered.vcf.gz) --target "$INDS" > ./treat_heterozygosity/CONROD.het

# STRCON
samples=$(cat popmap_files/STRCON.txt)
INDS=$(bcftools query -l ./filtered.vcf.gz | awk -v samples="$samples" '
BEGIN { split(samples, arr, "\n"); for (i in arr) sample_map[arr[i]] }
  { if ($0 in sample_map) printf "%s,", NR-1 }' | sed 's/,$//')
popStats -y GT --file <(bcftools view --threads 40 ./filtered.vcf.gz) --target "$INDS" > ./treat_heterozygosity/STRCON.het

# STRROD
samples=$(cat popmap_files/STRROD.txt)
INDS=$(bcftools query -l ./filtered.vcf.gz | awk -v samples="$samples" '
BEGIN { split(samples, arr, "\n"); for (i in arr) sample_map[arr[i]] }
  { if ($0 in sample_map) printf "%s,", NR-1 }' | sed 's/,$//')
popStats -y GT --file <(bcftools view --threads 40 ./filtered.vcf.gz) --target "$INDS" > ./treat_heterozygosity/STRROD.het

```

#Test
```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/
INDS=$(seq -s, 0 19)
popStats -y GT --file <( bcftools view --threads 40 ./filtered.vcf.gz) -t $INDS > ./treat_heterozygosity/test.het
```

#Add column with treatement file for each .het file
```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/

# all treatments
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tCONCON"}' ./treat_heterozygosity/CONCON.het) <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tSTRCON"}' ./treat_heterozygosity/STRCON.het) <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tCONROD"}' ./treat_heterozygosity/CONROD.het) <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tSTRROD"}' ./treat_heterozygosity/STRROD.het) > ./treat_heterozygosity/treatment.het

# CON-CON
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tCONCON"}' ./treat_heterozygosity/CONCON.het) > ./treat_heterozygosity/CONCON-treat.het

# STR-CON
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tSTRCON"}' ./treat_heterozygosity/STRCON.het) > ./treat_heterozygosity/STRCON-treat.het

# CON-ROD
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tCONROD"}' ./treat_heterozygosity/CONROD.het) > ./treat_heterozygosity/CONROD-treat.het

# STR-ROD
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tSTRROD"}' ./treat_heterozygosity/STRROD.het) > ./treat_heterozygosity/STRROD-treat.het
```

```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/treat_heterozygosity
mkdir -p mean_treat_het

bedtools merge -i <(mawk '!/CHR/' CONCON-treat.het| sort -k1,1 -k2,2n ) -c 4,5,6 -d -1 -o mean,mean,first > ../treat_heterozygosity/mean_treat_het/mean.CONCON.het
bedtools merge -i <(mawk '!/CHR/' STRCON-treat.het| sort -k1,1 -k2,2n ) -c 4,5,6 -d -1 -o mean,mean,first > ../treat_heterozygosity/mean_treat_het/mean.STRCON.het
bedtools merge -i <(mawk '!/CHR/' CONROD-treat.het| sort -k1,1 -k2,2n ) -c 4,5,6 -d -1 -o mean,mean,first > ../treat_heterozygosity/mean_treat_het/mean.CONROD.het
bedtools merge -i <(mawk '!/CHR/' STRROD-treat.het| sort -k1,1 -k2,2n ) -c 4,5,6 -d -1 -o mean,mean,first > ../treat_heterozygosity/mean_treat_het/mean.STRROD.het
```

```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/treat_heterozygosity/mean_treat_het/ 

#writing out tables in .txt files with headers

# all treatments
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tTREAT") mean.CONCON.het mean.STRCON.het mean.CONROD.het mean.STRROD.het > mean.treatment.het

cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.CONCON.het > mean.CONCON.het.txt
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.STRCON.het > mean.STRCON.het.txt
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.CONROD.het > mean.CONROD.het.txt
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.STRROD.het > mean.STRROD.het.txt

cd /home/ccaillon/ROD_CADO_CC_working/treat_heterozygosity/mean_treat_het/ 
mkdir -p ./Output/Heterozygosity
cp mean.treatment.het /home/ccaillon/ROD_CADO_CC_working/Output/Heterozygosity

cd /home/ccaillon/ROD_CADO_CC_working/treat_heterozygosity/
cp treatment.het /home/ccaillon/ROD_CADO_CC_working/Output/Heterozygosity
```

```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/Output/Heterozygosity/
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") treatment.het > treatment.het.txt
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.treatment.het > mean.treatment.het.txt
```

```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/Output/Heterozygosity/
#gzip treatment.het.txt
cp treatment.het.txt.gz /home/ccaillon/ROD_CADO/Diversity_metrics/Heterozygosity/
```

## Table
```{r}
setwd("~/ROD_CADO_CC_working/Output/Heterozygosity")

#read table into R
het.df<-read.table(paste("treatment.het", sep=""), sep="\t", header=T)
summary(het.df)
#set up dataframe for ggplot
het.df$TREAT <- factor(het.df$TREAT, levels=c("CONCON", "STRCON", "CONROD", "STRROD"))

#read table into R 
mean.het.df<-read.table(paste("mean.treatment.het", sep=""), sep="\t", header=T)
#set up df for ggplot
mean.het.df$TREAT <- factor(mean.het.df$TREAT, levels=c("CONCON", "STRCON", "CONROD", "STRROD"))
summary(mean.het.df)
```

# treatment.het
CHROM            BIN_START            BIN_END              N_SNPs             HET        
 NC_035784.1:12202592   Min.   :       36   Min.   :       37   Min.   :0.00000   Min.   :0.0000  
 NC_035782.1: 8074184   1st Qu.: 18450338   1st Qu.: 18450339   1st Qu.:0.04875   1st Qu.:0.0500  
 NC_035781.1: 7688972   Median : 36585803   Median : 36585804   Median :0.13875   Median :0.1500  
 NC_035780.1: 7418040   Mean   : 37805567   Mean   : 37805568   Mean   :0.19692   Mean   :0.2324  
 NC_035783.1: 7349464   3rd Qu.: 54560924   3rd Qu.: 54560925   3rd Qu.:0.34875   3rd Qu.:0.4000  
 NC_035787.1: 4655964   Max.   :104145289   Max.   :104145290   Max.   :0.50000   Max.   :1.0000  
 (Other)    :11134320                                                                             
    TREAT         
 CONCON:14630884  
 CONROD:14630884  
 STRCON:14630884  
 STRROD:14630884  

# mean.treatment.het
         CHROM            BIN_START            BIN_END             EXP_HET           OBS_HET      
 NC_035784.1:12202592   Min.   :       36   Min.   :       37   Min.   :0.00000   Min.   :0.0000  
 NC_035782.1: 8074184   1st Qu.: 18450338   1st Qu.: 18450339   1st Qu.:0.04875   1st Qu.:0.0500  
 NC_035781.1: 7688972   Median : 36585803   Median : 36585804   Median :0.13875   Median :0.1500  
 NC_035780.1: 7418040   Mean   : 37805567   Mean   : 37805568   Mean   :0.19692   Mean   :0.2324  
 NC_035783.1: 7349464   3rd Qu.: 54560924   3rd Qu.: 54560925   3rd Qu.:0.34875   3rd Qu.:0.4000  
 NC_035787.1: 4655964   Max.   :104145289   Max.   :104145290   Max.   :0.50000   Max.   :1.0000  
 (Other)    :11134320                                                                             
    TREAT         
 CONCON:14630884  
 STRCON:14630884  
 CONROD:14630884  
 STRROD:14630884  

## Plotting 
```{r}
library(ggplot2)
col_pal <- c("#0072B2", "#56B4E9", "#E69F00", "#F0E442")
# df$TREAT <- factor(df$TREAT, levels=c("CONCON", "STRCON", "CONROD", "STRROD"))
```

# HET by treatment 
```{r}
p1 <-ggplot(het.df, aes(x=HET,y = TREAT))+
  #geom_violin(aes(color=GROUP,fill= GROUP)) +
  #stat_summary(fun=mean, geom="point", shape=23, size=4)+
  #geom_dotplot(binaxis='y', stackdir='center', dotsize=0.005)+
  geom_boxplot(aes(fill=TREAT), width=0.75,outlier.shape = NA, outlier.color = "black")+
  stat_summary(fun=mean, geom="point", shape=23, size=2)+
  scale_fill_manual(values=col_pal)+
  scale_color_manual(values=col_pal)+
  scale_y_discrete(limits = rev(levels(het.df$TREAT)))+
  #ylab("Population/Line")+
  ggtitle("Heterozygosity")+
  coord_cartesian(xlim=c(0,0.5))+
  theme_classic() +
  labs(x="Heterozygosity")
p1
```

# Mean HET by treatment 
```{r}
p2 <-ggplot(mean.het.df, aes(x=OBS_HET,y = TREAT))+
  #geom_violin(aes(color=GROUP,fill= GROUP)) +
  #stat_summary(fun=mean, geom="point", shape=23, size=4)+
  #geom_dotplot(binaxis='y', stackdir='center', dotsize=0.005)+
  geom_boxplot(aes(fill=TREAT), width=0.75,outlier.shape = NA, outlier.color = "black")+
  stat_summary(fun=mean, geom="point", shape=23, size=2)+
  scale_fill_manual(values=col_pal)+
  scale_color_manual(values=col_pal)+
  scale_y_discrete(limits = rev(levels(mean.het.df$TREAT)))+
  #ylab("Population/Line")+
  ggtitle("Heterozygosity")+
  coord_cartesian(xlim=c(0,0.5))+
  theme_classic() +
  labs(x="Heterozygosity")
p2
```

## Statistics

# Summary
```{r}
library(dplyr)
library(plotrix)

group_by(het.df, TREAT) %>%
  dplyr::summarise(
    count = n(),
    mean = mean(HET, na.rm = TRUE),
    sd = sd(HET, na.rm = TRUE),
    se=std.error(HET, na.rm = TRUE) )
```

# Anova
```{r}
aov <- aov(HET ~ TREAT, het.df)
summary(aov)

TukeyHSD(aov, conf.level = 0.95)
```






