---
title: "Heterozygosity"
author: "Coline Caillon"
date: '2025-04-30'
output: github_document

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Jon's code: 
https://github.com/The-Eastern-Oyster-Genome-Project/2024_Eastern_Oyster_Population_Genomics/blob/main/Oyster_Genome_Population_Genomic_Analysis.md#heterozygosity

# Set up working directory and create symbolic link to the compressed VCF file

```{bash}
cd /home/ccaillon/
mkdir ROD_CADO_CC_working
cd ROD_CADO_CC_working
ln -s /home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz filtered.vcf.gz
```

```{bash}
# create popmap file with sample and treatment names
cd /home/ccaillon/ROD_CADO_CC_working
mkdir popmap_files
cd popmap_files
cp /home/Shared_Data/ROD_CADO/analysis/popmap treat_popmap
#add in treatment names to popmap file (w/ code)
head treat_popmap 
```

#Create treatment specific files containing a single column of all of the sample names within that treatment
```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/popmap_files

#Rename treatments
awk '$2 ~ /^C[123]$/ { $2 = "CONCON" } 
$2 ~ /^R[123]$/ { $2 = "CONROD" }
$2 ~ /^CH[123]$/ { $2 = "STRCON" }
$2 ~ /^RH[123]$/ { $2 = "STRROD" }
{ print }' treat_popmap > treat_popmap_renamed

awk '$2 == "CONCON" {print $1}' treat_popmap_renamed > CONCON.txt
awk '$2 == "STRCON" {print $1}' treat_popmap_renamed > STRCON.txt
awk '$2 == "CONROD" {print $1}' treat_popmap_renamed > CONROD.txt
awk '$2 == "STRROD" {print $1}' treat_popmap_renamed > STRROD.txt
```

#Subset the VCF by treatment and calculate heterozygosity for each treatment
```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/
mkdir -p treat_heterozygosity  # -p prevents error if directory exists

# CONCON
samples=$(cat popmap_files/CONCON.txt)
INDS=$(bcftools query -l ./filtered.vcf.gz | awk -v samples="$samples" '
BEGIN { split(samples, arr, "\n"); for (i in arr) sample_map[arr[i]] }
  { if ($0 in sample_map) printf "%s,", NR-1 }' | sed 's/,$//')
popStats -y GT --file <(bcftools view --threads 40 ./filtered.vcf.gz) --target "$INDS" > ./treat_heterozygosity/CONCON.het

# CONROD
samples=$(cat popmap_files/CONROD.txt)
INDS=$(bcftools query -l ./filtered.vcf.gz | awk -v samples="$samples" '
BEGIN { split(samples, arr, "\n"); for (i in arr) sample_map[arr[i]] }
  { if ($0 in sample_map) printf "%s,", NR-1 }' | sed 's/,$//')
popStats -y GT --file <(bcftools view --threads 40 ./filtered.vcf.gz) --target "$INDS" > ./treat_heterozygosity/CONROD.het

# STRCON
samples=$(cat popmap_files/STRCON.txt)
INDS=$(bcftools query -l ./filtered.vcf.gz | awk -v samples="$samples" '
BEGIN { split(samples, arr, "\n"); for (i in arr) sample_map[arr[i]] }
  { if ($0 in sample_map) printf "%s,", NR-1 }' | sed 's/,$//')
popStats -y GT --file <(bcftools view --threads 40 ./filtered.vcf.gz) --target "$INDS" > ./treat_heterozygosity/STRCON.het

# STRROD
samples=$(cat popmap_files/STRROD.txt)
INDS=$(bcftools query -l ./filtered.vcf.gz | awk -v samples="$samples" '
BEGIN { split(samples, arr, "\n"); for (i in arr) sample_map[arr[i]] }
  { if ($0 in sample_map) printf "%s,", NR-1 }' | sed 's/,$//')
popStats -y GT --file <(bcftools view --threads 40 ./filtered.vcf.gz) --target "$INDS" > ./treat_heterozygosity/STRROD.het

```

#Test
```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/
INDS=$(seq -s, 0 19)
popStats -y GT --file <( bcftools view --threads 40 ./filtered.vcf.gz) -t $INDS > ./treat_heterozygosity/test.het
```

#Add column with treatement file for each .het file
```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/

# all treatments
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tCONCON"}' ./treat_heterozygosity/CONCON.het) <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tSTRCON"}' ./treat_heterozygosity/STRCON.het) <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tCONROD"}' ./treat_heterozygosity/CONROD.het) <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tSTRROD"}' ./treat_heterozygosity/STRROD.het) > ./treat_heterozygosity/treatment.het

# CON-CON
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tCONCON"}' ./treat_heterozygosity/CONCON.het) > ./treat_heterozygosity/CONCON-treat.het

# STR-CON
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tSTRCON"}' ./treat_heterozygosity/STRCON.het) > ./treat_heterozygosity/STRCON-treat.het

# CON-ROD
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tCONROD"}' ./treat_heterozygosity/CONROD.het) > ./treat_heterozygosity/CONROD-treat.het

# STR-ROD
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tN_SITES\tN_SNPs\tHET\tTREAT") <(mawk '{if (NR>1) print $1 "\t" $2-1 "\t" $2 "\t" $4 "\t" $5 "\tSTRROD"}' ./treat_heterozygosity/STRROD.het) > ./treat_heterozygosity/STRROD-treat.het
```

```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/treat_heterozygosity
mkdir -p mean_treat_het

bedtools merge -i <(mawk '!/CHR/' CONCON-treat.het| sort -k1,1 -k2,2n ) -c 4,5,6 -d -1 -o mean,mean,first > ../treat_heterozygosity/mean_treat_het/mean.CONCON.het
bedtools merge -i <(mawk '!/CHR/' STRCON-treat.het| sort -k1,1 -k2,2n ) -c 4,5,6 -d -1 -o mean,mean,first > ../treat_heterozygosity/mean_treat_het/mean.STRCON.het
bedtools merge -i <(mawk '!/CHR/' CONROD-treat.het| sort -k1,1 -k2,2n ) -c 4,5,6 -d -1 -o mean,mean,first > ../treat_heterozygosity/mean_treat_het/mean.CONROD.het
bedtools merge -i <(mawk '!/CHR/' STRROD-treat.het| sort -k1,1 -k2,2n ) -c 4,5,6 -d -1 -o mean,mean,first > ../treat_heterozygosity/mean_treat_het/mean.STRROD.het
```

```{bash}
cd /home/ccaillon/ROD_CADO_CC_working/treat_heterozygosity/mean_treat_het/ 

#writing out tables in .txt files with headers

cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tTREAT") mean.CONCON.het mean.STRCON.het mean.CONROD.het mean.STRROD.het > mean.treatment.het

cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.CONCON.het > mean.CONCON.het.txt
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.STRCON.het > mean.STRCON.het.txt
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.CONROD.het > mean.CONROD.het.txt
cat <(echo -e "CHROM\tBIN_START\tBIN_END\tEXP_HET\tOBS_HET\tGROUP") mean.STRROD.het > mean.STRROD.het.txt

cd /home/ccaillon/ROD_CADO_CC_working/
mkdir -p ./Output/Heterozygosity
cp mean.treatment.het ./Output/Heterozygosity

```

## Table
```{r}
#read table into R
het.df<-read.table(paste("treatment.tajima", sep=""), sep="\t", header=T)
summary(tajd.df)
#set up dataframe for ggplot
tajd.df$TREAT <- factor(tajd.df$TREAT, levels=c("CONCON", "STRCON", "CONROD", "STRROD"))

#read table into R 
mean.tajd.df<-read.table(paste("mean.treatment.tajima", sep=""), sep="\t", header=T)
#set up df for ggplot
mean.tajd.df$TREAT <- factor(mean.tajd.df$TREAT, levels=c("CONCON", "STRCON", "CONROD", "STRROD"))
summary(mean.tajd.df)
```

## Plotting 
```{r}
library(ggplot2)
col_pal <- c("#0072B2", "#56B4E9", "#E69F00", "#F0E442")
# df$TREAT <- factor(df$TREAT, levels=c("CONCON", "STRCON", "CONROD", "STRROD"))
```






# Load requiered libraries

library(vcfR)   # for reading VCF files
library(dplyr)  # for data manipulation

# Set paths
vcf_path <- "/home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz"
output <- "~/ROD_CADO/Diversity_metrics/Heterozygosity_cc"  # directory to save results

# Read VCF file
vcf <- read.vcfR(vcf_path)





