---
title: "PCA analysis"
output:
  html_notebook:
    fig_caption: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    toc: yes
    toc_depth: '3'
  html_document:
    keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load library and convert vcf to pcadapt format
```{r}
#Load library
library(pcadapt)
# Use vcf2pcadapt to convert file type
vcf2pcadapt("/home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.sub1.vcf", output = "/home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/tmp.pcadapt", allele.sep = c("/", "|"))
```

# Set path to file and load data to filename variable

```{r}
#Set path to file
path_to_file <- "/home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/tmp.pcadapt"
#Load file to filename variable
filename <- read.pcadapt(path_to_file, type = "pcadapt")
```

# Examine PCs
```{r}
res <- pcadapt(filename, K = 5)
plot(res, option = "screeplot")
```

# Add popnames
```{r}
table <- read.table("/home/Shared_Data/ROD_CADO/analysis/popmap", header = FALSE)
poplist.names <- table[,2]
poplist.names

plot(res, option = "scores", pop = poplist.names)
```

# Analyze loadings
```{r}
par(mfrow = c(2, 2))
for (i in 1:4)
  plot(res$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
```

