---
title: "PCA analysis"
output:
  html_notebook:
    fig_caption: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    toc: yes
    toc_depth: '3'
  html_document:
    keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load library and convert vcf to pcadapt format
```{r}
#Load library
library(pcadapt)
```

# Set path to file and load data to filename variable

```{r}
#Set path to file
path_to_file <- "/home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/TRSdp10g1.FIL.bed"
#Load file to filename variable
filename <- read.pcadapt(path_to_file, type = "bed")
```

# Examine PCs
```{r}
res <- pcadapt(filename, K = 5)
plot(res, option = "screeplot")
```

# Add popnames
```{r}
table <- read.table("/home/jgreen/ROD_CADO_working/Nucleotide_diversity/popmap_files/treat_popmap", header = FALSE)
poplist.names <- table[,2]
poplist.names

plot(res, option = "scores", pop = poplist.names)
```

# Analyze loadings
```{r}
par(mfrow = c(2, 2))
for (i in 1:4)
  plot(res$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
```

# Investigate PC1

```{r}
plot(res$loadings[, 1], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
```

# Dealing with Linkage using LD clumping

```{r}
resld <- pcadapt(filename, K = 10, LD.clumping = list(size = 500, thr = 0.2))
plot(resld, option = "screeplot")
```

## Investigate loadings within the context for LD

```{r}
resld <- pcadapt(filename, K = 2, LD.clumping = list(size = 500, thr = 0.2))
par(mfrow = c(1, 1))
for (i in 1:2)
  plot(resld$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
```


## Plot PCA

```{r}
plot(resld, option = "scores", pop = poplist.names)
```

Main groupings that we see are seemingly group by family and not by treatment. We might not see treatment groupings looking at all SNPS, but may potentially see treatment groupings as we begin to select for outlier loci. 

### Manhattan Plot


```{r}
plot(res , option = "manhattan")
plot(resld , option = "manhattan")
```

### Q-Q plot

```{r}
plot(res, option = "qqplot")
plot(resld, option= "qqplot")
```

### Histograms of the test statistic and of the p-values

```{r}
hist(res$pvalues, xlab = "p-values", main = NULL, breaks = 50, col = "blue")
hist(resld$pvalues, xlab = "p-values", main = NULL, breaks = 50, col = "blue")
```

### Stats distribution

```{r}
plot(res, option = "stat.distribution")
plot(resld, option = "stat.distribution")
```