---
title: "Heterozygosity"
author: "Coline Caillon"
date: '2025-04-30'
output: github_document

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Jon's code: 
https://github.com/The-Eastern-Oyster-Genome-Project/2024_Eastern_Oyster_Population_Genomics/blob/main/Oyster_Genome_Population_Genomic_Analysis.md#heterozygosity

# Create symbolic link (shortcut) to a compressed VCF file

```{bash}
ln -s /home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz /home/ccaillon/ROD_CADO/Diversity_metrics/Heterozygosity_cc
```



## Plotting 
```{r}
library(ggplot2)
col_pal <- c("#0072B2", "#56B4E9", "#E69F00", "#F0E442")
# df$TREAT <- factor(df$TREAT, levels=c("CONCON", "STRCON", "CONROD", "STRROD"))
```






# Load requiered libraries

library(vcfR)   # for reading VCF files
library(dplyr)  # for data manipulation

# Set paths
vcf_path <- "/home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz"
output <- "~/ROD_CADO/Diversity_metrics/Heterozygosity_cc"  # directory to save results

# Read VCF file
vcf <- read.vcfR(vcf_path)





