---
title: "ROD CADO Outlier Exploration"
author: "Amy Zyck"
date: '2025-08-18'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# ROD CADO Outlier Exploration

Looking for overlapping outlier SNPs identified through LFMM, pcadapt, and Outflank:

Bed file locations:

[LFMM2](https://github.com/MarineEvoEcoLab/ROD_CADO/blob/main/Outlier_detection/outliers_lfmm2.bed)
[pcadapt](https://github.com/MarineEvoEcoLab/ROD_CADO/blob/main/PCA_analysis/outlier_pcadapt_thinned_alpha01.loci.txt)
[OutFlank](https://github.com/MarineEvoEcoLab/ROD_CADO/blob/main/FST_Outflank.bed)

Upload libraries

```{r}
library(vcfR)
library(tibble)
library(ggplot2)
library(pcadapt)
library(tidyverse)
library(dplyr)
library(qvalue)
library(vcfR)
```

**Converting pcadapt txt file to bed format**

```{bash}
awk '{print $1 "\t" $2-1 "\t" $2}' outlier_pcadapt_thinned_alpha01.loci.txt | bedtools sort -i - > outliers_pcadapt.bed
```

**Extracting Outflank outliers from bed file**

Jon had made a bed file for all the OutFlank SNPs. Located on Github. The bed file was a Git LFS pointer file, which is basically a way to make larger files in the repo smaller for your local copy. To access the larger version of the file, I did:

```{bash}
git lfs install
git lfs pull
```

This file contains the chromosome, bp positions, Fst calculations, and designations of outlier SNPs (TRUE v FALSE) for each SNP in the analysis. I want to extract the SNPs that are specifically outliers (TRUE), and save the chromosome and bp position columns to a new file:

```{bash}
awk '$5=="TRUE"{print $1"\t"$2"\t"$3}' FST_Outflank.bed > Outflank_outliers.bed
```

Now it matches the format of the LFMM and pcadapt outlier bed files. 


Looking at outliers that are overlapped across the three detection programs:

```{bash}
# First sort each bed file so they're all in the same order
sort Outflank_outliers.bed > outliers_outflank_sorted.txt
sort outliers_lfmm2.bed > outliers_lfmm2_sorted.txt
sort outliers_pcadapt.bed > outliers_pcadapt_sorted.txt
```

```{bash}
comm -12 outliers_pcadapt_sorted.txt outliers_lfmm2_sorted.txt | wc -l
```

Three shared outlier SNPs between pcadapt and LFMM2.

```{bash}
comm -12 outliers_pcadapt_sorted.txt outliers_lfmm2_sorted.txt
```

```{bash}
comm -12 outliers_outflank_sorted.txt outliers_lfmm2_sorted.txt | wc -l
```

No shared outlier SNPs between outflank and LFMM2.

```{bash}
comm -12 outliers_pcadapt_sorted.txt outliers_outflank_sorted.txt | wc -l
```

Five shared outlier SNPs between pcadapt and outflank.

```{bash}
comm -12 outliers_pcadapt_sorted.txt outliers_outflank_sorted.txt
```

## Convert Bed file using vcftools and the original VCF file

```{bash}
#ln -s /home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz .
vcftools --gzvcf SNP.TRSdp10g1.FIL.vcf.gz --bed outliers_lfmm2.bed --recode --recode-INFO-all --out outliers_lfmm2.recode.bed
```



## PCA plot of LFMM outliers

Load bed files into objects

```{r}
path_to_file <- "/home/jgreen/ROD_CADO/Outlier_Exploration/outliers_lfmm2.recode.bed.recode.vcf"
lfmm_snps <- read.pcadapt(path_to_file, type = "vcf")
```

```{r}
res_lfmm <- pcadapt(lfmm_snps, K = 5)
plot(res_lfmm, option = "screeplot")
```

# Add popnames
```{r}
table <- read.table("/home/jgreen/ROD_CADO_working/Nucleotide_diversity/popmap_files/treat_popmap", header = FALSE)
poplist.names <- table[,2]
poplist.names

plot(res_lfmm, option = "scores", pop = poplist.names)
```

# Analyze loadings
```{r}
par(mfrow = c(3, 2))
for (i in 1:5)
  plot(res_lfmm$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
```

# Investigate PC1, PC2, and PC3

```{r}
plot(res_lfmm$loadings[, 1], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
plot(res_lfmm$loadings[, 2], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
plot(res_lfmm$loadings[, 3], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
```

## Plot PCA

```{r}
plot(res_lfmm, option = "scores", pop = poplist.names)
```


