---
title: "ROD CADO Outlier Exploration"
author: "Amy Zyck"
date: '2025-08-18'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# ROD CADO Outlier Exploration

Looking for overlapping outlier SNPs identified through LFMM, pcadapt, and Outflank:

Bed file locations:

[LFMM2](https://github.com/MarineEvoEcoLab/ROD_CADO/blob/main/Outlier_detection/outliers_lfmm2.bed)
[pcadapt](https://github.com/MarineEvoEcoLab/ROD_CADO/blob/main/PCA_analysis/outlier_pcadapt_thinned_alpha01.loci.txt)
[OutFlank](https://github.com/MarineEvoEcoLab/ROD_CADO/blob/main/FST_Outflank.bed)

Upload libraries

```{r}
library(vcfR)
library(tibble)
library(ggplot2)
library(pcadapt)
library(tidyverse)
library(dplyr)
library(qvalue)
library(vcfR)
library(lfmm) 
#library(LEA)
library(vegan) 
library(adegenet)
```

**Converting pcadapt txt file to bed format**

```{bash}
awk '{print $1 "\t" $2-1 "\t" $2}' outlier_pcadapt_thinned_alpha01.loci.txt | bedtools sort -i - > outliers_pcadapt.bed
```

**Extracting Outflank outliers from bed file**

Jon had made a bed file for all the OutFlank SNPs. Located on Github. The bed file was a Git LFS pointer file, which is basically a way to make larger files in the repo smaller for your local copy. To access the larger version of the file, I did:

```{bash}
git lfs install
git lfs pull
```

This file contains the chromosome, bp positions, Fst calculations, and designations of outlier SNPs (TRUE v FALSE) for each SNP in the analysis. I want to extract the SNPs that are specifically outliers (TRUE), and save the chromosome and bp position columns to a new file:

```{bash}
awk '$5=="TRUE"{print $1"\t"$2"\t"$3}' FST_Outflank.bed > Outflank_outliers.bed
```

Now it matches the format of the LFMM and pcadapt outlier bed files. 


Looking at outliers that are overlapped across the three detection programs:

```{bash}
# First sort each bed file so they're all in the same order
sort Outflank_outliers.bed > outliers_outflank_sorted.txt
sort outliers_lfmm2.bed > outliers_lfmm2_sorted.txt
sort outliers_pcadapt.bed > outliers_pcadapt_sorted.txt
```

```{bash}
comm -12 outliers_pcadapt_sorted.txt outliers_lfmm2_sorted.txt | wc -l
```

Three shared outlier SNPs between pcadapt and LFMM2.

```{bash}
comm -12 outliers_pcadapt_sorted.txt outliers_lfmm2_sorted.txt
```

```{bash}
comm -12 outliers_outflank_sorted.txt outliers_lfmm2_sorted.txt | wc -l
```

No shared outlier SNPs between outflank and LFMM2.

```{bash}
comm -12 outliers_pcadapt_sorted.txt outliers_outflank_sorted.txt | wc -l
```

Five shared outlier SNPs between pcadapt and outflank.

```{bash}
comm -12 outliers_pcadapt_sorted.txt outliers_outflank_sorted.txt
```

## Convert Bed file using vcftools and the original VCF file

```{bash}
#ln -s /home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz .
vcftools --gzvcf SNP.TRSdp10g1.FIL.vcf.gz --bed outliers_lfmm2.bed --recode --recode-INFO-all --out outliers_lfmm2.recode.bed
```
```{r}
vcf_out_lfmm_C_ADD <- read.vcfR("outliers_lfmm2.recode.bed.recode.vcf")
```

## Setting up strata_ROD and other objects

```{r}
# Setting up strata file - population info and environmental data 
strata_ROD <- read.table("strata_ROD_tab", header=TRUE)
```

```{r}
pop <- c(rep("C", 20),rep("CH", 20),rep("R", 20),rep("RH", 20))
```

```{r}
Stress <- c(rep("No", 20),rep("Yes", 20),rep("No", 20),rep("Yes", 20))

Disease <- c(rep("No", 20),rep("No", 20),rep("Yes", 20),rep("Yes", 20))
```


```{r}
rad_out_lfmm_C_ADD.filt <- vcfR2genind(vcf_out_lfmm_C_ADD, strata = strata_ROD, pop = c(rep("C", 20),rep("CH", 20),rep("R", 20),rep("RH", 20)))

rad_out_lfmm_C_ADD.filt
```

```{r}
# Plot a PCA for the all outlier loci
x_out_lfmm_C_ADD <- scaleGen(rad_out_lfmm_C_ADD.filt, NA.method = "mean")
out_lfmm_C_ADD_pca <- dudi.pca(x_out_lfmm_C_ADD, center = FALSE, scale = FALSE, scannf = FALSE, nf = 4)
eig_percent_out_lfmm_C_ADD <- round((out_lfmm_C_ADD_pca$eig / (sum(out_lfmm_C_ADD_pca$eig))) * 100, 2)

pca_dat_out_lfmm_C_ADD <- as.data.frame(out_lfmm_C_ADD_pca$li) %>%
                    rownames_to_column(var = "ind") %>%
                    cbind(pop = pop(rad_out_lfmm_C_ADD.filt)) %>%
                    mutate(pop = factor(pop, levels = c("C", "CH","R", "RH")))

pca_dat_out_lfmm_C_ADD <- pca_dat_out_lfmm_C_ADD %>%
  mutate(Combined = ifelse(pop %in% c("RH"), "Yes", "No"))
```

```{r}
pc_scores_C_ADD <- out_lfmm_C_ADD_pca$li
```

```{r}
permanova_result_C_ADD <- adonis2(pc_scores_C_ADD ~ pop, method = "euclidean")
print(permanova_result_C_ADD)
```

```{r}
permanova_result_C_ADD <- adonis2(pc_scores_C_ADD ~ Combined, method = "euclidean")
print(permanova_result_C_ADD)
```

```{r}
pca_data_C_ADD <- data.frame(PC1 = pc_scores_C_ADD[,1], PC2 = pc_scores_C_ADD[,2], Treatment = pop)

# Define new treatment names
treatment_names <- c("C" = "CONCON",
                         "CH" = "STRCON",
                         "R" = "CONROD",
                         "RH" = "STRROD")

# Define custom colors
custom_colors <- c("C" = "#0072B2", "CH" = "#56B4E9", "R" = "#E69F00", "RH" = "#F0E442")

# Extract p-value from PERMANOVA result
p_value_C_ADD <- permanova_result_C_ADD$`Pr(>F)`[1]

eigenvalues_C_ADD <- eigen(cov(pca_data_C_ADD[, c("PC1", "PC2")]))$values

# Calculate proportion of variance explained
pve_C_ADD <- eigenvalues_C_ADD / sum(eigenvalues_C_ADD) * 100
```

```{r}
# Create the plot with proportion of variance on axes
p <- ggplot(pca_data_C_ADD, aes(x = PC1, y = PC2, color = Treatment)) +
  geom_point() +
  stat_ellipse() +
  theme_bw() +
  scale_color_manual(values = custom_colors, labels = treatment_names) +
  xlab(paste0("PC1 (", round(pve_C_ADD[1], 2), "%)")) +
  ylab(paste0("PC2 (", round(pve_C_ADD[2], 2), "%)")) +
  ggtitle("PCA Plot of LFMM SNPs with Treatment Groups") +
  annotate("text", x = -Inf, y = -Inf, 
           label = paste("PERMANOVA\n p =", format(p_value_C_ADD, digits = 2)),
           hjust = -0.1, vjust = -1)

ggsave("pca_plot_LFMM_outliers.png", width = 8, height = 6, dpi = 300)
```

```{r}
# Enhanced version with better formatting and options
p <- ggplot(pca_data_C_ADD, aes(x = PC1, y = PC2, color = Treatment)) +
  geom_point(size = 3, alpha = 0.7) +                    # Larger, semi-transparent points
  stat_ellipse(level = 0.95, linetype = 2) +             # 95% confidence ellipses, dashed
  theme_minimal() +
  theme(
    legend.position = "right",
    panel.grid.minor = element_blank(),                   # Remove minor grid lines
    plot.title = element_text(hjust = 0.5)               # Center title
  ) +
  scale_color_manual(
    values = custom_colors, 
    labels = treatment_names,
    name = "Treatment"                                    # Legend title
  ) +
  xlab(paste0("PC1 (", round(pve_C_ADD[1], 1), "% variance explained)")) +
  ylab(paste0("PC2 (", round(pve_C_ADD[2], 1), "% variance explained)")) +
  ggtitle("PCA Plot with Treatment Groups") +
  annotate("text", 
           x = -Inf, y = -Inf,
           label = paste0("PERMANOVA: p = ", 
                         ifelse(p_value_C_ADD < 0.001, "< 0.001", 
                                format(round(p_value_C_ADD, 3), nsmall = 3))),
           hjust = -0.1, vjust = -1.5,
           size = 3.5,
           fontface = "italic")

ggsave("pca_plot_LFMM_outliers.png", width = 8, height = 6, dpi = 300)
```

## PCA analysis of Outflank files

## Convert Bed file using vcftools and the original VCF file

```{bash}
#ln -s /home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz .
vcftools --gzvcf SNP.TRSdp10g1.FIL.vcf.gz --bed outliers_lfmm2.bed --recode --recode-INFO-all --out outliers_lfmm2.recode.bed
```

```{r}
vcf_out_outflank_C_ADD <- read.vcfR("outliers_Outflank.recode.bed.recode.vcf")
```

## Setting up strata_ROD and other objects

```{r}
# Setting up strata file - population info and environmental data 
strata_ROD <- read.table("strata_ROD_tab", header=TRUE)
```

```{r}
pop <- c(rep("C", 20),rep("CH", 20),rep("R", 20),rep("RH", 20))
```

```{r}
Stress <- c(rep("No", 20),rep("Yes", 20),rep("No", 20),rep("Yes", 20))

Disease <- c(rep("No", 20),rep("No", 20),rep("Yes", 20),rep("Yes", 20))
```


```{r}
rad_out_outflank_C_ADD.filt <- vcfR2genind(vcf_out_Outflank_C_ADD, strata = strata_ROD, pop = c(rep("C", 20),rep("CH", 20),rep("R", 20),rep("RH", 20)))

rad_out_outflank_C_ADD.filt
```

```{r}
# Plot a PCA for the all outlier loci
x_out_outflank_C_ADD <- scaleGen(rad_out_outflank_C_ADD.filt, NA.method = "mean")
out_outflank_C_ADD_pca <- dudi.pca(x_out_outflank_C_ADD, center = FALSE, scale = FALSE, scannf = FALSE, nf = 4)
eig_percent_out_outflank_C_ADD <- round((out_outflank_C_ADD_pca$eig / (sum(out_outflank_C_ADD_pca$eig))) * 100, 2)

pca_dat_out_outflank_C_ADD <- as.data.frame(out_outflank_C_ADD_pca$li) %>%
                    rownames_to_column(var = "ind") %>%
                    cbind(pop = pop(rad_out_outflank_C_ADD.filt)) %>%
                    mutate(pop = factor(pop, levels = c("C", "CH","R", "RH")))

pca_dat_out_outflank_C_ADD <- pca_dat_out_outflank_C_ADD %>%
  mutate(Combined = ifelse(pop %in% c("RH"), "Yes", "No"))
```

```{r}
pc_scores_C_ADD <- out_outflank_C_ADD_pca$li
```

```{r}
permanova_result_C_ADD <- adonis2(pc_scores_C_ADD ~ pop, method = "euclidean")
print(permanova_result_C_ADD)
```

```{r}
permanova_result_C_ADD <- adonis2(pc_scores_C_ADD ~ Combined, method = "euclidean")
print(permanova_result_C_ADD)
```

```{r}
pca_data_C_ADD <- data.frame(PC1 = pc_scores_C_ADD[,1], PC2 = pc_scores_C_ADD[,2], Treatment = pop)

# Define new treatment names
treatment_names <- c("C" = "CONCON",
                         "CH" = "STRCON",
                         "R" = "CONROD",
                         "RH" = "STRROD")

# Define custom colors
custom_colors <- c("C" = "#0072B2", "CH" = "#56B4E9", "R" = "#E69F00", "RH" = "#F0E442")

# Extract p-value from PERMANOVA result
p_value_C_ADD <- permanova_result_C_ADD$`Pr(>F)`[1]

eigenvalues_C_ADD <- eigen(cov(pca_data_C_ADD[, c("PC1", "PC2")]))$values

# Calculate proportion of variance explained
pve_C_ADD <- eigenvalues_C_ADD / sum(eigenvalues_C_ADD) * 100
```

```{r}
# Create the plot with proportion of variance on axes
p <- ggplot(pca_data_C_ADD, aes(x = PC1, y = PC2, color = Treatment)) +
  geom_point() +
  stat_ellipse() +
  theme_bw() +
  scale_color_manual(values = custom_colors, labels = treatment_names) +
  xlab(paste0("PC1 (", round(pve_C_ADD[1], 2), "%)")) +
  ylab(paste0("PC2 (", round(pve_C_ADD[2], 2), "%)")) +
  ggtitle("PCA Plot of outflank SNPs with Treatment Groups") +
  annotate("text", x = -Inf, y = -Inf, 
           label = paste("PERMANOVA\n p =", format(p_value_C_ADD, digits = 2)),
           hjust = -0.1, vjust = -1)

ggsave("pca_plot_outflank_outliers.png", width = 8, height = 6, dpi = 300)
```

```{r}
# Enhanced version with better formatting and options
p <- ggplot(pca_data_C_ADD, aes(x = PC1, y = PC2, color = Treatment)) +
  geom_point(size = 3, alpha = 0.7) +                    # Larger, semi-transparent points
  stat_ellipse(level = 0.95, linetype = 2) +             # 95% confidence ellipses, dashed
  theme_minimal() +
  theme(
    legend.position = "right",
    panel.grid.minor = element_blank(),                   # Remove minor grid lines
    plot.title = element_text(hjust = 0.5)               # Center title
  ) +
  scale_color_manual(
    values = custom_colors, 
    labels = treatment_names,
    name = "Treatment"                                    # Legend title
  ) +
  xlab(paste0("PC1 (", round(pve_C_ADD[1], 1), "% variance explained)")) +
  ylab(paste0("PC2 (", round(pve_C_ADD[2], 1), "% variance explained)")) +
  ggtitle("PCA Plot with Treatment Groups") +
  annotate("text", 
           x = -Inf, y = -Inf,
           label = paste0("PERMANOVA: p = ", 
                         ifelse(p_value_C_ADD < 0.001, "< 0.001", 
                                format(round(p_value_C_ADD, 3), nsmall = 3))),
           hjust = -0.1, vjust = -1.5,
           size = 3.5,
           fontface = "italic")

ggsave("pca_plot_outflank_outliers.png", width = 8, height = 6, dpi = 300)
```