---
title: "Genetic Diversity work"
output:
  html_notebook:
    fig_caption: yes
    toc: yes
    toc_depth: 3
    toc_float: yes
  pdf_document:
    toc: yes
    toc_depth: '3'
  html_document:
    keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, root_dir = "/home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered")
```

# Load libraries and file paths
```{r, message=FALSE, warning=FALSE}
library("plyr")
library("dplyr")
library("ggplot2")
library(R.utils)
library(gghighlight)
library(ggman)
library(ggtext)
library(patchwork)
library(plotrix)
library(qqman)
library(qvalue)
library(reshape2)
library(tidyr)
library(zoo)
library(infer)
options(dplyr.summarise.inform = FALSE)
library(bigsnpr)
library("wesanderson")
library("directlabels")
library(OutFLANK)
library(adegenet)
library(poppr)
library(vcfR)
library(stringr)
library(matrixStats)
```


# Nucleotide diversity

Nucleotide diversity (often referred to using the symbol π) is the average pairwise difference between all possible pairs of individuals in your sample. It is a very intuitive and simple measure of genetic diversity, and is accurately estimated even with very few samples. A formal definition is here.

We can obtain the nucleotide diversity (π) from our VCF file using vcftools software. In our case we will collect the π value from each 10 kb (10,000 bp) window of the genome.

NB: vcftools is a very flexible tool for analyzing, manipulating VCF files. It can do many other wonderful things. The vcftools manual is on github here (https://vcftools.sourceforge.net/man_latest.html).

## Start of modified workflow

## Setup 

The following was run on the command line
```{bash}
# Make ROD_CADO_working directory in home
mkdir ROD_CADO_working
cd ROD_CADO_working
# Make Nucleotide_diversity directory
mkdir ROD_CADO_working
# create popmap file with sample and treatment names
cp /home/Shared_Data/ROD_CADO/analysis/popmap popmap
# manually add in treatment names to popmap file (w/ code)
head treat_popmap 
```
C1_2    con-con
C1_4    con-con
C1_5    con-con
C1_6    con-con
C1_7    con-con
C1_8    con-con
C1_9    con-con
C2_10   con-con
C2_11   con-con
C2_2    con-con

Create treatment specific files containing a single column of all of the sample names within that treatment
```{bash}
awk '$2 == "con-con" {print $1}' treat_popmap > con-con.txt | awk '$2 == "str-con" {print $1}' treat_popmap > str-con.txt | awk '$2 == "con-rod" {print $1}' treat_popmap > con-rod.txt | awk '$2 == "str-rod" {print $1}' treat_popmap > str-rod.txt
```

### Run VCF tools PI window
```{bash eval = FALSE}
#bcftools view --threads 20 -S SNP.TRSdp10g1.FIL.vcf | vcftools --vcf -  --window-pi 10000 --out ROD.CADO.all.pi

# For con-con
# Step 1: Filter VCF for population subset
vcftools --gzvcf SNP.TRSdp10g1.FIL.vcf.gz --keep con-con.txt --recode --recode-INFO-all --out temp_concon_filtered

# Step 2: bgzip output
bgzip temp_concon_filtered.recode.vcf

# Step 3: Calculate windowed pi
vcftools --gzvcf temp_concon.filtered.recode.vcf.gz --window-pi 10000 --out ROD.CADO.con-con.pi.windowed.pi

# For str-con
# Step 1: Filter VCF for population subset
vcftools --gzvcf SNP.TRSdp10g1.FIL.vcf.gz --keep popmap_files/str-con.txt --recode --recode-INFO-all --out temp_strcon_filtered

# Step 2: bgzip output
bgzip temp_strcon_filtered.recode.vcf

# Step 3: # Step 3: Calculate windowed pi
vcftools --gzvcf temp_strcon_filtered.recode.vcf.gz --window-pi 10000 --out ROD.CADO.str-con.pi.windowed.pi

# For con-rod
# Step 1: Filter VCF for population subset
vcftools --gzvcf SNP.TRSdp10g1.FIL.vcf --keep popmap_files/con-rod.txt --recode --recode-INFO-all --out temp_conrod_filtered

# Step 2: bgzip output
bgzip temp_conrod_filtered.recode.vcf

# Step 3: Calculate windowed pi
vcftools --gzvcf temp_conrod_filtered.recode.vcf.gz --window-pi 10000 --out ROD.CADO.con-rod.pi.windowed.pi

# For str-rod
# Step 1: Filter VCF for population subset
vcftools --gzvcf SNP.TRSdp10g1.FIL.vcf --keep popmap_files/str-rod.txt --recode --recode-INFO-all --out temp_strrod_filtered

# Step 2: bgzip output
bgzip temp_strrod_filtered.recode.vcf

# Step 3: Calculate windowed pi
vcftools --gzvcf temp_strrod_filtered.recode.vcf.gz --window-pi 10000 --out ROD.CADO.str-rod.pi.windowed.pi
```

### Make loop with help from ChatGPT

#### Make script

```{bash}
#!/bin/bash

VCF=SNP.TRSdp10g1.FIL.vcf.gz
POPS=("con-con" "str-con" "con-rod" "str-rod")

for POP in "${POPS[@]}"; do
    echo "Processing $POP..."

    KEEP="popmap_files/${POP}.txt"
    OUT_PREFIX="temp_${POP//-}"
    REC_VCF="${OUT_PREFIX}.recode.vcf"
    REC_VCFGZ="${REC_VCF}.gz"
    OUTPUT_PI="ROD.CADO.${POP}.pi.windowed.pi"

    # Step 1: Filter and recode
    vcftools --gzvcf "$VCF" \
        --keep "$KEEP" \
        --recode --recode-INFO-all \
        --out "$OUT_PREFIX"

    # Step 2: Compress VCF and remove uncompressed
    bgzip "$REC_VCF"
    rm "$REC_VCF"

    # Step 3: Calculate windowed pi
    vcftools --gzvcf "$REC_VCFGZ" \
        --window-pi 10000 \
        --out "$OUTPUT_PI"

    # Step 4: Clean up compressed VCF
    rm "$REC_VCFGZ"

    echo "Finished processing $POP"
    echo "---------------------------"
done
```

#### Make executable

```{bash}
chmod +x run_pi_calculations.sh
```

#### Run in tmux
```{bash}
# Run in tmux
tmux new -s pi_calc
# Reattach later
tmux attach-session -t pi_calc
```
### Load dataframe
```{r, message=FALSE, warning=FALSE}
pi.all.dataframe<-read.table("/home/Shared_Data/ROD_CADO/analysis/raw.vcf/ROD.CADO.all.pi.windowed.pi", sep="\t", header=T)
```

### Color palette

```{r}
#Here is the color pallette that we will use for everything:

col_pal <- c("#0072B2", "#56B4E9", "#E69F00", "#F0E442")

#Let's factor treatments as follows:

df$TREAT <- factor(df$TREAT, levels=c("CONCON", "STRCON", "CONROD", "STRROD"))
```

### Modify CHROM column in dataframe
```{r, message=FALSE, warning=FALSE}
pi.all.dataframe %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035780.1", "1")) %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035781.1", "2")) %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035782.1", "3")) %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035783.1", "4")) %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035784.1", "5")) %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035785.1", "6")) %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035786.1", "7")) %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035787.1", "8")) %>%
  mutate(CHROM = str_replace(CHROM, "NC_035788.1", "9")) %>% 
  mutate(CHROM = str_replace(CHROM, "NC_035789.1", "10"))  -> mydf
mydf$CHROM <- as.factor(mydf$CHROM)
```


### Descriptive statistics
```{r}
summary(mydf)
by(mydf, mydf$CHROM, summary)
cor(mydf$N_VARIANTS, mydf$PI)
```
### Plot PI by chromosome
```{r, message=FALSE, warning=FALSE}
ggplot(mydf, aes(x=CHROM, y=PI,))+
  geom_violin(aes(color=CHROM,fill=CHROM))+
  geom_boxplot(aes(fill=CHROM), width=0.1,outlier.shape = 23, outlier.color = "black")+
  stat_summary(fun=mean, geom="point", shape=23, size=2)+
  scale_fill_brewer(palette = "Paired")+
  theme_classic()
```

### Smaller visualizations
```{r, message=FALSE, warning=FALSE}
hist(mydf$PI,br=40)

boxplot(mydf$PI, ylab="Nuc Diversity")
```

### Plot By position
```{r, message=FALSE, warning=FALSE}
ggplot(mydf, aes(x=BIN_START, y=PI, color=CHROM))+
  geom_point()+
  facet_wrap(~CHROM)+
  theme_classic()
```

```{r, message=FALSE, warning=FALSE}
# Subset by chrom
mydf.chr1 <- mydf[which(mydf$CHROM=="1"),]

ggplot(mydf.chr1, aes(x=BIN_START, y=PI))+
  geom_point()+
  theme_classic()
```

```{r}
# Create an empty list to store the plots
plot_list <- list()

# Loop through chromosomes 1 to 10
for (i in 1:10) {
  chr_data <- mydf[which(mydf$CHROM == as.character(i)), ]
  
  p <- ggplot(chr_data, aes(x = BIN_START, y = PI)) +
    geom_point() +
    theme_classic() +
    ggtitle(paste("Chromosome", i))
  
  # Store the plot in the list
  plot_list[[paste0("chr", i)]] <- p
}

# To display one of the plots, for example:
print(plot_list[["chr3"]])

```

### Breaking up pi by treatment type?

I believe that the next step would be to compare nucleotide diversity between the different treatment groups. I need to think about how to pipe the different treatment groups from the VCF file to the PI command. Maybe I don't need to do this but can help with other metrics?

# Runs of homozygosity

Runs of homozygosity (ROH) are contiguous lengths of homozygous genotypes that are present in an individual due to parents transmitting identical haplotypes to their offspring.

The potential of predicting or estimating individual autozygosity for a subpopulation is the proportion of the autosomal genome above a specified length, termed Froh.

This technique can be used to identify the genomic footprint of inbreeding in conservation programs, as organisms that have undergone recent inbreeding will exhibit long runs of homozygosity. The effect of inbreeding in the resulting sub-populations could be studied by measuring the runs of homozygosity in different individuals.

## Start ROH workflow

```{bash}
vcftools --vcf SNP.TRSdp10g1.FIL.vcf --LROH --out ROD.CADO.all.LROH
```

