---
title: "Tajimas_D"
author: "Megan Guidry"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## About Tajima's D
Tajima's D is a normalized version of observed avg. pairwise difference amoung individuals across the genome minus the expected avg. pairwise difference amoung individuals across the genome.

Negative = selection removing variation or recent population expansion; 
Positive = selection maintaining variation or a recent population contraction.

[video explanation](https://www.youtube.com/watch?v=wiyay4YMq2A)

Jon's code: 
https://github.com/The-Eastern-Oyster-Genome-Project/2024_Eastern_Oyster_Population_Genomics/blob/main/Oyster_Genome_Population_Genomic_Analysis.md#tajimas-d

## Setup 
on the command line, activate conda environment and set up files 
```{bash}
conda create -n vk python=3.8 numpy scipy
conda activate vk
pip install VCF-kit


#set up files on github repo
cd /home/mguidry/ROD_CADO/Diversity_metrics/
mkdir Tajimas_D
nano TD_analysis.md

#set up working directory on KITT and link files
cd /home/mguidry/ROD_CADO_MG_working/
mkdir Tajimas_D
cd Tajimas_D
ln -s /home/Shared_Data/ROD_CADO/analysis/raw.vcf/filtered/SNP.TRSdp10g1.FIL.vcf.gz filtered.vcf.gz
```

## [VCF-kit `tajima`](https://vcf-kit.readthedocs.io/en/latest/tajima/ )
Bins are useful for investigating particular areas of the genome that could be regions of interest. 
Sliding windows are helpful for scanning the entire genome.

The following was run on the command line with tmux
```{bash}
#calculated tajimas d in bins of 10,000 bp
vk tajima 10000 10000 filtered.vcf.gz > 10kb_bin_td_filtered_vcf.tajima 

#can also do a sliding window approach 
vk tajima 100000 1000 filtered.vcf.gz > window_td_filtered_vcf.tajima

#or a continuous sliding window approach to capture every unique bin of variants that fall within 100kb of one another
##did not explore this yet as we are just looking for broad trends right now and have lots of data 
#vk tajima 100000 --sliding filtered.vcf.gz > sliding_window_td_filtered_vcf.tajima
```

## Assign reads to different treatments
Need metadata to do this!

Chat about which treatments to compare
- 2 treatments (ROD vs control)
- 4 treatemtns (stress-ROD, ambient-ROD, stress-control, ambient-control)
```{bash}

```


## Prep data for R
```{r}
#binned data (10,000 base bins)
bin.td.df<-read.table(paste("10kb_bin_td_filtered_vcf.tajima", sep=""), sep="\t", header=T)
summary(bin.td.df)

#windowed data (100kb bins that slide 1000 bases each iterations)
win.td.df<-read.table(paste("window_td_filtered_vcf.tajima", sep=""), sep="\t", header=T)
summary(win.td.df)
```

## Plot

example ggplot from Jon's code
```{r}
bd <-ggplot(tajd.wvd.dataframe, aes(x=TAJD,y = TREAT))+
  #geom_violin(aes(color=GROUP,fill= GROUP)) +
  #stat_summary(fun=mean, geom="point", shape=23, size=4)+
  #geom_dotplot(binaxis='y', stackdir='center', dotsize=0.005)+
  geom_boxplot(aes(fill=GROUP), width=0.75,outlier.shape = 23, outlier.color = "black")+
  stat_summary(fun=mean, geom="point", shape=23, size=2)+
  scale_fill_manual(values=c("#D55E00", "#009E73"))+
  scale_color_manual(values=c("#D55E00", "#009E73"))+
  scale_y_discrete(limits = rev(levels(tajd.wvd.dataframe$POP)))+
  #ylab("Population/Line")+
  #ggtitle("Wild vs. Selected Tajima's *D*")+
  theme_classic() +
  labs(x="Tajima's *D*") +theme(axis.title.x = element_markdown(),  axis.title.y = element_blank(), plot.title = element_markdown()) 

bd <- bd + guides(color = "none", fill ="none")
```




